src/runtime/hvx_128.ll

```diff
@@ -131,3 +131,114 @@ define weak_odr <64 x i32> @halide.hexagon.mul.vw.vuh(<64 x i32> %a, <64 x i16>
   %ab = call <64 x i32> @llvm.hexagon.V6.vcombine.128B(<32 x i32> %ab_hi, <32 x i32> %ab_lo)
   ret <64 x i32> %ab
 }
+
+; Hexagon is missing shifts for byte sized operands.
+declare <32 x i32> @llvm.hexagon.V6.vaslh.128B(<32 x i32>, i32)
+declare <32 x i32> @llvm.hexagon.V6.vasrh.128B(<32 x i32>, i32)
+declare <32 x i32> @llvm.hexagon.V6.vlsrh.128B(<32 x i32>, i32)
+declare <32 x i32> @llvm.hexagon.V6.vaslhv.128B(<32 x i32>, <32 x i32>)
+declare <32 x i32> @llvm.hexagon.V6.vasrhv.128B(<32 x i32>, <32 x i32>)
+declare <32 x i32> @llvm.hexagon.V6.vlsrhv.128B(<32 x i32>, <32 x i32>)
+declare <64 x i32> @llvm.hexagon.V6.vzb.128B(<32 x i32>)
+declare <64 x i32> @llvm.hexagon.V6.vsb.128B(<32 x i32>)
+declare <32 x i32> @llvm.hexagon.V6.vshuffeb.128B(<32 x i32>, <32 x i32>)
+
+define weak_odr <128 x i8> @halide.hexagon.shl.vub.ub(<128 x i8> %a, i8 %b) nounwind uwtable readnone alwaysinline {
+  %a_32 = bitcast <128 x i8> %a to <32 x i32>
+  %bw = zext i8 %b to i32
+  %aw = call <64 x i32> @llvm.hexagon.V6.vzb.128B(<32 x i32> %a_32)
+  %aw_lo = call <32 x i32> @llvm.hexagon.V6.lo.128B(<64 x i32> %aw)
+  %sw_lo = call <32 x i32> @llvm.hexagon.V6.vaslh.128B(<32 x i32> %aw_lo, i32 %bw)
```

Comment:
The shift and the or can be done in one instruction on the scalar side. But I think you are right. The current implementation takes only 4 instructions (I am not counting the "hi" and "lo" because they should disappear after register allocation.